### PR TITLE
Some issues fixed.

### PR DESCRIPTION
**Pull request by sept-en**
Hello!

I've fixed some issues and added some minor improvements:
- Fixed [issue #11469](https://hub.qgis.org/issues/14469) and also the searching algorithm was optimized because before it took too much time;
- Fixed [issue #12580](https://hub.qgis.org/issues/12580);
- ParameterMultipleInput: added tests;
- A little of PEP8 fixes.

Best regards,
Kirill Abramov.

### CODE REVIEW

File: python/plugins/processing/core/parameters.py [hunk 1]
```diff
@@ -338,6 +341,38 @@ def __init__(self, name='', description='', datatype=-1, optional=False):
...
+    def setMinNumInputs(self, _minNumInputs):
+        if _minNumInputs is None:
+            self.minNumInputs = 0
+            return True
+
+        if _minNumInputs < 1 and not self.optional:
+            # dont allow to set negative or null number of inputs if parameter isn't optional
+            return False
+
+        self.minNumInputs = int(minNumInputs)
```
Comment (alexbruy) [inline python/plugins/processing/core/parameters.py]: Typo? There is no `minNumInputs` variable

**Comment by alexbruy** [comment]
Also indentation test failed on some commits

File: python/plugins/processing/core/parameters.py (hunk 1 above)
Comment (sept-en) [inline python/plugins/processing/core/parameters.py]: Oh, yes, just a typo, I'm sorry, my bad.

File: python/plugins/processing/core/parameters.py [hunk 2]
```diff
@@ -350,9 +385,14 @@ def setValue(self, obj):
         if isinstance(obj, list):
             if len(obj) == 0:
                 if self.optional:
+                    self.value = None
                     return True
                 else:
                     return False
+            # prevent setting value if we didn't provide required minimal number of inputs
+            elif len(obj) < minNumInputs:
```
Comment (alexbruy) [inline python/plugins/processing/core/parameters.py]: Looks like typo too. Shouldn't it be `self.minNumInputs`?

Comment (sept-en) [inline python/plugins/processing/core/parameters.py]: Oh, shame on me.
Fixed.

**Pull request merged**